### ISSUE TITLE
How to speed up the training？

### ISSUE THREAD

**Issue by DavidLining** [Unity-Technologies/ml-agents#835]
I am training a hundred drones to fly by themselves with algorithm PPO. 
All of drones are using same Brain, I found the training is so slow that I have to wait a long time to verify the result after changing hyper-parameters.
I don't know what make the training become slowly. Maybe the brain using 312  Observations and it's difficult for unity3D to render environment with high frequency under the configuration of my computer. 
During the training, the CPU usage is about 50% and the RAM usage is about 50%.  It seems that the training have not use out the CPU and RAM resource. So I doubt that the render speed makes the traing slowly. ??
**Does anyone know how to speed up the training ?**  It difficult for me to upgrade hardware configuration right now, so i need to find if there are other ways to speed up the training.

Brain Parameters:
  Vector Observation: 
       Space Type: Continuous
       Space Size: 312( the positions of one hundred drones, the position of destination, the bounds of 
                                 space,  the velocity of object ) 
       Stacked Vectors: 3
  Vector Action:
       Space Type: Continuous
       Space Size: 4 (Directions on x, y, z axis and speed)
My computer configuration:
  CPU: i5-4310M @2.7GHz
  RAM: 8.00GB
  SSD: 128GB

**Comment by MarcoMeter** [comment]
One limitation is that the Unity build cannot run as fast as possible (timescale limit). However, this shouldn't affect you too much. You are stacking 312 inputs 3 times. So you feed 936 values to the neural net. That's quite expensive. It also depends on how big your neural net is. How many hidden layers and nodes are you using?

It sounds like that you are working on a really big environment, so long training times seem to be normal.

**Comment by DavidLining** [comment]
@MarcoMeter I using 2 hidden layers， I am not clear about the nodes, i have not found the parameter to set the nodes.  The timescale is 100, but it hard to view the running of Unity build. It seems blocked!

**Comment by MarcoMeter** [comment]
By nodes I mean hidden_units.

The "blocked behavior" is due to how the communication between the Unity build and Python works. Both applications are not communicating each frame.

**Comment by DavidLining** [comment]
@MarcoMeter  oh, haha, the hidden_units is 128.  So the "blocked behavior" will not affect the training speed, right?

**Comment by MarcoMeter** [comment]
It is meant to work like this.

My first suggestion is to scale down your model. Reduce or aggregate your state space.
Also, I wouldn't train 100 agents at once. Start with like 6 or 10.

**Comment by DavidLining** [comment]
@MarcoMeter  thanks! Let me ask you a final question. If I start with less agents and get the model, how to train the 100 agents based on the old model? 
I start to learn the deep learning recently so that not very clear about it.

**Comment by MarcoMeter** [comment]
Are all agents part of one environment?

Regardless, if you are just getting started, start out small. Simplify your environment and once progress kicks in raise the complexity step by step.

**Comment by DavidLining** [comment]
@MarcoMeter Yes, all agents are part of environment. All agents must avoid collision during the way to the destination. I want to simulate and implement the design like following video: 
[https://www.youtube.com/watch?v=VnTQTm7vNbY&t=8s](url)

**Comment by eshvk** [comment]
One thing you could do to speed up things is run Unity by turning off graphics and running it in batch mode. This works in your situation because you are using visual observations. You'd have to modify `environment.py` to do something like:

```
                    self.proc1 = subprocess.Popen(
                        [launch_string,'-nographics', '-batchmode',
                         '--port', str(self.port)])
```

**Comment by DavidLining** [comment]
@eshvk Great, graphics will not show during training. what 'batchmode' used for?

**Comment by xiaomaogy** [comment]
Thanks for reaching out to us. Hopefully you were able to resolve your issue.  We are closing this due to inactivity, but if you need additional assistance, feel free to reopen the issue.
